community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCacheTest.java

```diff
@@ -156,28 +161,127 @@ public void shouldAddGroupInTheMiddleIfTwo() throws Exception
+                    random.nextInt( 1_000_000 ), true );
+            assertEquals( 1, cache.getCount( i, i % 5, directions[i % directions.length] ) );
+        }
+
+        // WHEN
+        cache.clearRelationships();
+
+        // THEN
+        for ( int i = 0; i < nodes; i++ )
+        {
+            assertEquals( -1L, cache.getFirstRel( nodes, groupVisitor ) );
+            assertEquals( 1, cache.getCount( i, i % 5, directions[i % directions.length] ) );
+        }
+    }
+
+    @Test
+    public void shouldGetAndPutRelationshipAroundChunkEdge() throws Exception
+    {
+        // GIVEN
+        cache = new NodeRelationshipCache( NumberArrayFactory.HEAP, 10 );
+
+        // WHEN
+        long nodeId = 1_000_000 - 1;
+        int type = 0;
+        Direction direction = Direction.OUTGOING;
+        long relId = 10;
+        cache.getAndPutRelationship( nodeId, type, direction, relId, false );
+
+        // THEN
+        assertEquals( relId, cache.getFirstRel( nodeId, mock( GroupVisitor.class ) ) );
+    }
+
+    @Test
+    public void shouldPutRandomStuff() throws Exception
+    {
+        // GIVEN
+        cache = new NodeRelationshipCache( NumberArrayFactory.HEAP, 10, 1000, base );
+
+        // WHEN
+        for ( int i = 0; i < 10_000; i++ )
+        {
+            cache.getAndPutRelationship( random.nextInt( 100_000 ), random.nextInt( 5 ),
+                    Direction.OUTGOING, random.nextInt( 1_000_000 ), true );
+        }
+    }
+
+    @Test
+    public void shouldPut6ByteRelationshipIds() throws Exception
+    {
+        // GIVEN
+        cache = new NodeRelationshipCache( NumberArrayFactory.HEAP, 1, 100, base );
+        long sparseNode = 0;
+        long denseNode = 1;
+        long relationshipId = (1L << 48) - 2;
+        cache.incrementCount( denseNode );
+
+        // WHEN
+        assertEquals( -1L, cache.getAndPutRelationship( sparseNode, 0, OUTGOING, relationshipId, false ) );
+        assertEquals( -1L, cache.getAndPutRelationship( denseNode, 0, OUTGOING, relationshipId, false ) );
+
+        // THEN
+        GroupVisitor groupVisitor = mock( GroupVisitor.class );
+        assertEquals( relationshipId, cache.getAndPutRelationship( sparseNode, 0, OUTGOING, 1, false ) );
+        assertEquals( relationshipId, cache.getAndPutRelationship( denseNode, 0, OUTGOING, 1, false ) );
+    }
+
+    @Test
+    public void shouldFailFastIfTooBigRelationshipId() throws Exception
+    {
+        // GIVEN
+        cache = new NodeRelationshipCache( NumberArrayFactory.HEAP, 1, 100, base );
+
+        // WHEN
+        cache.getAndPutRelationship( 0, 0, OUTGOING, (1L << 48) - 2, false );
+        try
+        {
+            cache.getAndPutRelationship( 0, 0, OUTGOING, (1L << 48) - 1, false );
+            fail( "Should fail" );
+        }
+        catch ( IllegalArgumentException e )
+        {
+            // THEN Good
+            assertTrue( e.getMessage().contains( "max" ) );
```

Comment:
`@Test( exception = IllegalArgumentException.class )`? Don't like in general since you can't assert exception message and you can't assert that the exception is thrown from exactly the place you expect it to.